app/lua/lrodefs.h

```diff
@@ -38,10 +37,17 @@
   return 1
 #endif
 
-#define LROT_TABENTRY(n,t)  {LSTRKEY(#n), LRO_ROVAL(t)}
-#define LROT_FUNCENTRY(n,f) {LSTRKEY(#n), LRO_FUNCVAL(f)}
-#define LROT_NUMENTRY(n,x)  {LSTRKEY(#n), LRO_NUMVAL(x)}
-#define LROT_END            {LNILKEY, LNILVAL}
+#define LROT_TABLE(t)        static const LUA_REG_TYPE t ## _map[];
+#define LROT_TABLEREF(t)     ((void *) t ## _map)
+#define LROT_BEGIN(t)        static const LUA_REG_TYPE t ## _map [] = {
+#define LROT_PUBLIC_BEGIN(t) const LUA_REG_TYPE t ## _map[] = {
+#define LROT_EXTERN(t)       extern const LUA_REG_TYPE t ## _map[]
+#define LROT_TABENTRY(n,t)   {LRO_STRKEY(#n), LRO_ROVAL(t ## _map)},
+#define LROT_FUNCENTRY(n,f)  {LRO_STRKEY(#n), LRO_FUNCVAL(f)},
+#define LROT_NUMENTRY(n,x)   {LRO_STRKEY(#n), LRO_NUMVAL(x)},
+#define LROT_LUDENTRY(n,x)   {LRO_STRKEY(#n), LRO_LUDATA((void *) x)},
+#define LROT_END(t,mt, f)    {LRO_NILKEY, LRO_NILVAL} };
```

Comment:
Future proofing.  A ROTable is currently a `luaR_entry` vector, so the Table code need to include conditional logic `(is the address in Flash) ? do a ROTable reference : do a Table reference`  On the next version the `LROT_END()` will generate a `luaR_entry` terminator followed by a readonly `Table` rec which points to the vector.  Generating the extra fields in the module source now means that we can implement this change without needing to change the modules again.

The main advantage of ROTables having a `Table` record is that we can push the RO vs RW abstraction down a level an dump a lot of conditional testing in the Lua VM where it only needs to reference and manipulate the table as an entity.  The Table record contains useful stuff like the size of the table, and flags to indicate whether metas such as `__index`, `_newindex` and  `__gc` exist.

The [ROTable lookup cache](https://github.com/nodemcu/nodemcu-firmware/blob/aa2e29d3dd6deb982df52da05834441558668f2a/app/lua/lrotable.c#L54] now give a better than 95% hit rate for key-found searches of ROTables.  (I've tweaked the lookup algo to give more uniform coverage of the cache and to improve performance, BTW.)  There is still a residual performance issue with key misses (even though this code runs about 4× faster than the eLua version).  Having the abitity use the flabs byte to short-circuit frequent meta misses, and the length to allow you to sort the key list into ascending order and tag that you can do a binary search on the longer ROTables such as the baselib one, will pretty much  remove this penalty all together.
    So that in future we can re